### PR TITLE
Remove course (phase out a course) 

### PR DESCRIPTION
**Pull request by AlexLam1997**

### CODE REVIEW

File: backend/logic/courses/courses.js [hunk 1]
```diff
@@ -145,9 +147,26 @@ var addCourseOfferings = async courseOfferings => {
   }
 };
```
Comment (zerfuth) [inline backend/logic/courses/courses.js]: Please specify an API description for the function
```suggestion
/**
 * Phases out a course that is no longer offered
 * @author Alex Lam
 * @param {string} courseCode
 * @returns true if successful
 * @throws error if MySQL connection failed
 *         invalid format course code if course code format is incorrect
 *         false if modification failed
*/
```

File: backend/logic/courses/courses.js [hunk 2]
```diff
@@ -145,9 +147,26 @@ var addCourseOfferings = async courseOfferings => {
...
 
+let phaseOutCourse = async courseCode => {
```
Comment (zerfuth) [inline backend/logic/courses/courses.js]: Make sure you do a validation check for the course code.

File: backend/logic/courses/courses.js (hunk 1 above)
Comment (AlexLam1997) [reply backend/logic/courses/courses.js]: I put it with the endpoint in the other file

File: backend/logic/courses/courses.js (hunk 2 above)
Comment (AlexLam1997) [reply backend/logic/courses/courses.js]: true, thanks!

File: backend/logic/courses/courses.js (hunk 1 above)
Comment (zerfuth) [reply backend/logic/courses/courses.js]: yeah but from i thought we were doing an API description for the endpoint and also one for the actual methods?

File: backend/logic/courses/courses.js [hunk 3]
```diff
@@ -145,9 +147,29 @@ var addCourseOfferings = async courseOfferings => {
...
+  let connection = await mysql.getNewConnection();
+  try {
+    if (!courseCode) {
```
Comment (zerfuth) [inline backend/logic/courses/courses.js]: check format of course code? format.js might already have a function for that, otherwise could write a simple one

**Comment by Ty-Won** [comment]
great job, would it be possible to change the pull request title? It's simply for commit history reasons so that when we look back if we need to reset, we at least know what this is all about

File: backend/logic/courses/courses.js (hunk 2 above)
Comment (Ty-Won) [reply backend/logic/courses/courses.js]: Try a regex expression, this is just a quick one but it should match 4 capital letters a single space and 3 digits 0-9
use https://regex101.com/ to help you if you do regex
```regex
courseCode.match(/[A-Z]{4}[ ]{1}\d{3}/g)
```

File: backend/logic/courses/courses.js [hunk 4]
```diff
@@ -145,9 +147,27 @@ var addCourseOfferings = async courseOfferings => {
   }
 };
 
+let phaseOutCourse = async courseCode => {
+  let connection = await mysql.getNewConnection();
+  try {
+    await format.verifyCourse(courseCode);
+    await connection.query(
+      "UPDATE courses SET phased_out = TRUE WHERE course_code = ?",
```
Comment (Ty-Won) [inline backend/logic/courses/courses.js]: Not 100% sure, but from stevens explanation from the slack 

> @channel I don't think I explained myself clear during our meeting for beginning a transaction in MySQL. There is absolutely no need to begin and commit a transaction unless if you're writing data into the database (UPDATE, DELETE, INSERT, etc.). By default, MySQL database have auto commits turned on, but, in some cases, you want to temporarily disable auto commits until you're sure a block of queries will execute successfully which is where transactions come in. A transaction is used to preserve atomicity on multiple queries meaning that if one of the queries fails then the entire transaction must be rollback 

TLDR; If you're working with SELECTs only, you don't need to begin a transaction since it prevents legitimate write instructions from accessing the table

Do you have to use the transaction and commit function?

File: backend/logic/courses/courses.js (hunk 2 above)
Comment (zerfuth) [reply backend/logic/courses/courses.js]: also, i think someone made a function for that (at least i saw it on @GarethDPeters's branch, so you could just reuse that.

File: backend/logic/courses/courses.js (hunk 4 above)
Comment (si-x-li) [reply backend/logic/courses/courses.js]: A race condition is possible where connection A removes data before connection B updates it but this a likely extremely rare race condition considering how often courses get phased out (maybe once every 5 years) and either way if the call fails there's no net impact on the database so it's more for styling points lol

File: backend/logic/courses/courses.js (hunk 3 above)
Comment (si-x-li) [reply backend/logic/courses/courses.js]: You can use **verifyCourseCode()** which just got merged into dev

File: backend/logic/courses/courses.js (hunk 1 above)
Comment (AlexLam1997) [reply backend/logic/courses/courses.js]: whaat, its the same description though thats so redundant, I dont think we should do that

File: backend/logic/courses/courses.js (hunk 2 above)
Comment (AlexLam1997) [reply backend/logic/courses/courses.js]: done! thanks :D

File: backend/logic/courses/courses.js (hunk 1 above)
Comment (zerfuth) [reply backend/logic/courses/courses.js]: The route might be calling more than one function, each function should have a description of what it does. Just like javadocs man
At least that's my opinion on it, it's minor thing.

File: backend/logic/courses/courses.js (hunk 2 above)
Comment (AlexLam1997) [reply backend/logic/courses/courses.js]: done!

File: db/update scripts/1.03.sql [hunk 5]
```diff
@@ -0,0 +1,6 @@
+SET SQL_SAFE_UPDATES
```
Comment (si-x-li) [inline db/update scripts/1.03.sql]: Not a fan of turning off SQL_SAFE_UPDATES. Is there anyway to proceed without having this?

File: backend/logic/courses/courses.js [hunk 6]
```diff
@@ -261,12 +263,30 @@ var updateCourse = async (course, newTitle, newTags) => {
...
+      courseCode
+    );
+    return true;
+  } catch (error) {
+    console.error(error);
+    throw error;
```
Comment (si-x-li) [inline backend/logic/courses/courses.js]: Not recommended to throw back SQL errors back at frontend since users will be able to see the structure of the database. Consider returning a custom message or masking the error instead.

File: backend/logic/courses/courses.js [hunk 7]
```diff
@@ -261,12 +263,30 @@ var updateCourse = async (course, newTitle, newTags) => {
   }
 };
 
+let phaseOutCourse = async courseCode => {
+  let connection = await mysql.getNewConnection();
+  try {
+    await format.verifyCourse(courseCode);
```
Comment (si-x-li) [inline backend/logic/courses/courses.js]: This can be moved outside of the try block since you're rethrowing the error anyways

File: db/update scripts/1.03.sql (hunk 5 above)
Comment (AlexLam1997) [reply db/update scripts/1.03.sql]: Not really since I need to update all the entries where phased out is currently null :/ No way to write it in terms of the course_code

File: backend/logic/courses/courses.js (hunk 6 above)
Comment (AlexLam1997) [reply backend/logic/courses/courses.js]: Youre right ill fix that thanks!

File: backend/logic/courses/courses.js (hunk 1 above)
Comment (AlexLam1997) [reply backend/logic/courses/courses.js]: Ahh I see what you mean. Alright ill add it thanks
